File: mobile/apps/photos/plugins/ente_photos_platform/android/src/main/kotlin/io/ente/photos/platform/processlock/ProcessLockRegistry.kt
package io.ente.photos.platform.processlock

import android.util.Log

object ProcessLockRegistry {
    private class Holder(
        val instanceId: String,
        var origin: String,
        var operation: String,
        val acquiredAtNanos: Long,
    )

    private const val TAG = "ProcessLockRegistry"

    private val guard = Any()
    private val holders = HashMap<String, Holder>()

    fun tryAcquire(
        name: String,
        instanceId: String,
        origin: String,
        operation: String,
    ): Boolean =
        synchronized(guard) {
            val current = holders[name]
            when {
                current == null -> {
                    holders[name] = Holder(instanceId, origin, operation, System.nanoTime())
                    true
                }
                // A Dart hot restart can leave this plugin instance holding the lock.
                current.instanceId == instanceId -> {
                    Log.i(
                        TAG,
                        "Healed same-instance holder of '$name' " +
                            "(${current.origin}/${current.operation} -> $origin/$operation)",
                    )
                    current.origin = origin
                    current.operation = operation
                    true
                }
                else -> false
            }
        }

    fun release(name: String, instanceId: String): Boolean =
        synchronized(guard) {
            if (holders[name]?.instanceId == instanceId) {
                holders.remove(name)
                true
            } else {
                false
            }
        }

    fun releaseAllForInstance(instanceId: String) {
        synchronized(guard) {
            val released = holders.filterValues { it.instanceId == instanceId }
            if (released.isNotEmpty()) {
                Log.w(
                    TAG,
                    "Engine detach released lock(s) still held: " +
                        released.entries.joinToString {
                            "${it.key} (${it.value.origin}/${it.value.operation})"
                        },
                )
            }
            holders.values.removeAll { it.instanceId == instanceId }
        }
    }

    fun state(name: String): Map<String, Any>? =
        synchronized(guard) {
            holders[name]?.let {
                mapOf(
                    "origin" to it.origin,
                    "operation" to it.operation,
                    "heldForMillis" to (System.nanoTime() - it.acquiredAtNanos) / 1_000_000,
                )
            }
        }
}

File: mobile/apps/photos/plugins/ente_photos_platform/ios/Sources/PhotosPlatformCore/ProcessLockRegistry.swift
import Foundation

final class ProcessLockRegistry: @unchecked Sendable {
    static let shared = ProcessLockRegistry()

    private struct Holder {
        let instanceId: String
        var origin: String
        var operation: String
        let acquiredAt: DispatchTime
    }

    private let lock = NSLock()
    private var holders: [String: Holder] = [:]

    private init() {}

    func tryAcquire(
        name: String,
        instanceId: String,
        origin: String,
        operation: String
    ) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        if var current = holders[name] {
            // A Dart hot restart can leave this plugin instance holding the lock.
            guard current.instanceId == instanceId else { return false }
            NSLog(
                "ProcessLockRegistry: healed same-instance holder of '%@' (%@/%@ -> %@/%@)",
                name, current.origin, current.operation, origin, operation
            )
            current.origin = origin
            current.operation = operation
            holders[name] = current
            return true
        }
        holders[name] = Holder(
            instanceId: instanceId,
            origin: origin,
            operation: operation,
            acquiredAt: .now()
        )
        return true
    }

    func release(name: String, instanceId: String) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        guard holders[name]?.instanceId == instanceId else { return false }
        holders.removeValue(forKey: name)
        return true
    }

    func releaseAll(instanceId: String) {
        lock.lock()
        defer { lock.unlock() }
        let released = holders.filter { $0.value.instanceId == instanceId }
        if !released.isEmpty {
            let description =
                released
                .map { "\($0.key) (\($0.value.origin)/\($0.value.operation))" }
                .joined(separator: ", ")
            NSLog(
                "ProcessLockRegistry: engine detach released lock(s) still held: %@",
                description
            )
        }
        holders = holders.filter { $0.value.instanceId != instanceId }
    }

    func state(name: String) -> [String: Any]? {
        lock.lock()
        defer { lock.unlock() }
        guard let holder = holders[name] else { return nil }
        let heldForNanos =
            DispatchTime.now().uptimeNanoseconds - holder.acquiredAt.uptimeNanoseconds
        return [
            "origin": holder.origin,
            "operation": holder.operation,
            "heldForMillis": Int64(heldForNanos / 1_000_000),
        ]
    }
}
